ddtrace/internal/module.py

```diff
@@ -189,7 +197,18 @@ def _create_module(self, spec):
         return self.loader.create_module(spec)
 
     def _exec_module(self, module):
-        self.loader.exec_module(module)
+        patched_source = None
+        if IS_IAST_ENABLED and _should_iast_patch(module.__name__):
```

Comment:
I'm reluctant to have product-specific logic into a generic module. Can we think of a different approach to this? I think this is (perhaps just partly) tackled in #4294

Reply:
Well, since this is called at import time, and we need to do the `exec_module` differently, I found this to be the most simple and explicit way to do it. I'm open to suggestions to do it differently, but think we'll end up patching all the customer's code and then some third-party modules as well.

Reply:
@P403n1x87 so you want for the IAST part to do a different `_ImportHookChainedLoader` on a different module? I think this would be harder to follow and debug in case of problems with imports. Having the patching code here is IMO easier to follow and being under an "if IS_IAST_ENABLED" (like all IAST code will be) I don't see the problem.

Reply:
I can see a dedicated `Loader`/`Finder` in `sys.meta_path` as a much cleaner approach to this. To me this looks more like a hack inside `ModuleWatchdog`, whose purpose is to provide module import hooks. But like I said in my previous comment, I would be happy if this change proposed an extended API for the `ModuleWatchdog` hooks whereby we can accommodate for AST patching prior to execution. Then IAST could use that API instead in its internal area.

As for the IAST guard, I don't see how we could add a guard for the top-level import
~~~
from ddtrace.appsec.iast._util import _is_iast_enabled
~~~
which would then be the potential source of circular imports in the future.

I'm still not entirely convinced the approach proposed here is the best we can go for.

Reply:
> I can see a dedicated Loader/Finder in sys.meta_path as a much cleaner approach to this.

I second this. This could be a good opportunity to expand the interface of the import hook mechanism in such a way that it could be usable by other products in the future.

> Having the patching code here is IMO easier to follow...I don't see the problem

The clearest reason that I see to avoid putting product-specific code in `internal` is the increased cognitive load that it would place on future editors of `internal/module.py`. When changing this file, our job is made easier by the fact that we don't have to think about specific products as long as certain guarantees are maintained. With this change, people wanting to change this file would have to understand how IAST fits into that picture in greater detail than they would otherwise. For an internal module that's depended on by multiple products, I think it's worthwhile to be very careful about increasing cognitive load.

Reply:
I've changed the approach to remove the product import from the internal module, but unsure if the new approach is the right one. Resolving this conversation for now